### ISSUE TITLE
[css-images-4] Color stop fixup: do interpolation hints function as positioned color stops?

### ISSUE THREAD

**Issue by danburzo** [w3c/csswg-drafts#3931]
I'm looking at [adding support](https://github.com/Evercoder/culori/pull/77) for color stop positions and interpolation hints to a color library I've been working on, and it's not clear for me from the spec: when [fixing up color stops](https://drafts.csswg.org/css-images-4/#color-stop-fixup), and we spread color stops missing a position position evenly between the color stops which have one:

> 3. If any __color stop__ still does not have a position, then, __for each run of adjacent color stops without positions__, set their positions so that they are __evenly spaced between the preceding and following color stops with positions__. 

...do the color hints count as _color stops_ in step 3 of the process?

Here's my library's interpretation (top) vs. browser interpretation (bottom) of the following gradient:

```css
linear-gradient(to right, blue, red, 50%, green)
```

<img width="371" alt="Screenshot 2019-05-14 at 15 32 28" src="https://user-images.githubusercontent.com/205375/57698121-82ccb280-765d-11e9-90b1-b81c0a9abb12.png">

Is it normal that by introducing a color hint between `red` and `green`, the position of `red` shifts from 0.5 to 0.25? My mental model for the process resembled Photoshop's behavior, where moving the midpoints between two stops would not alter the position of the stops.

<img width="423" alt="Screenshot 2019-05-14 at 15 46 37" src="https://user-images.githubusercontent.com/205375/57698970-8f520a80-765f-11e9-9701-99dd3628d855.png">

**Comment by danburzo** [comment]
A couple of other observations from the field:

__1. Boundary cases for logarithm function:__

I've made a [quick notebook](https://observablehq.com/@danburzo/css-gradient-interpolation-hints) to explore how the interpolation hint's position affects the interpolation curve. For `H = 0` and `H = 1` the logarithm gives (understandably) weird results and there might need to be some adjustments to the formula. This seems to work as expected:

```js
const C = (P, H) => 
  H === 0 ? 1 : 
  H === 1 ? 0 : 
  Math.pow(P, Math.log(0.5) / Math.log(H));
```

i.e. for `H = 0`, paint completely with next color stop, and for `H = 1` paint completely with previous color stop. (Also note that Photoshop limits H to the [0.05, 0.95] range to avoid the extreme curves)

__ 2. Interpolation hint fixup example__

The example below seems to suggest that there's a step 4 (possibly computation of midpoint color?), and suggests the midpoint is split into two linear gradients, which I don't think is the case anymore. (I'm also curious if that was the case in a previous iteration of the spec.)

```
8. linear-gradient(red, 25%, white)
   =14=>
   linear-gradient(red 0%, rgb(100%,50%,50%) 25%, white 100%) 
```

**Comment by danburzo** [comment]
One last thing: as I was writing [this comment](https://github.com/w3c/csswg-drafts/issues/1332#issuecomment-492555433), I realized that with this hypothetical `midpoint` function (and with the possible addition of easing functions in the syntax), browsers can still support their _current_ understanding of color stop position computation (which _I think_ is incorrect), and have a _decoupled_ midpoint function going forward.

**Comment by AmeliaBR** [comment]
I agree that this needs to be defined more explicitly.

Given that existing implementations seem to agree, and their behavior isn't illogical, I think the spec should match implementations: color hint positions should be factored in equally with color stop positions when distributing stops without explicit positions.

It is unfortunate that this makes it difficult to expand to a generalized easing function approach, but I think your suggestion in #3935 is a good way to get around that.

**Comment by cabanier** [comment]
> i.e. for H = 0, paint completely with next color stop, and for H = 1 paint completely with previous color stop. (Also note that Photoshop limits H to the [0.05, 0.95] interval to avoid the extreme curves.)

FYI all implementations in browsers of the color hint move H to be in that color range

**Comment by danburzo** [comment]
Ah, good to know! Also, if I interpreted the code & comments correctly, browsers will approximate the logarithm by a series of nine evenly-distributed color stops — six on the longer interval, and two on the shorter one, possibly to compensate for Skia missing a native way to render eased gradients (?)

**Comment by cabanier** [comment]
Yes, I believe I used up to 9 color stops. Most of the time, it would be less.
That is an implementation detail.
The platform libraries didn't support this idiom, so I approximated it.

Ideally, the drawing libraries are extended so they actually draw the exponential ramp

**Comment by cabanier** [comment]
FWIW interpolation hints are specified as regular color stops in the spec. They just lack a color.
I'm unsure what needs to be changed in the spec.

**Comment by AmeliaBR** [comment]
> FWIW interpolation hints are specified as regular color stops in the spec.

I'm not sure how you read it that way? Color hints and color stops are defined as separate syntax elements. They need to be, to prohibit starting or ending on a hint.  In [step 2 of the algorithm](https://drafts.csswg.org/css-images-4/#color-stop-fixup) there is a fix-up for positions that specifically applies to both hints and stops. But step 3 currently only applies to stops.

I'm also curious about your comment (https://github.com/w3c/csswg-drafts/issues/3931#issuecomment-494662385) about adjusting the hint to be within a reasonable range. There should probably be some informative text in the spec about that, even if we don't stipulate the exact bounds normatively.

**Comment by danburzo** [comment]
_With the disclaimer that I have no insight to the history of the feature and the iterations the implementations, and the spec, went through:_ 

I can relate to the "interpolation hints are regular color stops" angle if we assume that:
- an interpolation hint is meant to be the position between colors `A` and `B` of a computed average color `(A + B) / 2`, i.e. a _midpoint_ H.
- with the available graphics engines, the transition between `A` and `B` through `H` can be _at least_ two linear gradients (`A→H`, `H→B`), _acceptably_ a series of linear gradients that approximate the log function, and _ideally_ the log function implemented at the graphics engine level. 

However, solely by reading the current spec and proposed algorithm of color stop positioning, I did not get that impression at all and was surprised when comparing with browser implementations.

**Comment by cabanier** [comment]
> an interpolation hint is meant to be the position between colors A and B of a computed average color (A + B) / 2, i.e. a midpoint H.

That is correct and the spec clearly spells this out.

> with the available graphics engines, the transition between A and B through H can be at least two linear gradients (A→H, H→B), acceptably a series of linear gradients that approximate the log function, and ideally the log function implemented at the graphics engine level.

If the hint is in the middle, it's only a single linear gradient.
The approximation is an implementation detail. The spec doesn't need to call this out.

**Comment by cabanier** [comment]
> I'm not sure how you read it that way? Color hints and color stops are defined as separate syntax elements. They need to be, to prohibit starting or ending on a hint. In step 2 of the algorithm there is a fix-up for positions that specifically applies to both hints and stops. 

Browser run the fix-up step on the position of the stops without looking if they are hints or regular color stops.

> But step 3 currently only applies to stops.

I think that's because all hints should be normalized after step 2 so there's no need to call them out.

> I'm also curious about your comment (#3931 (comment)) about adjusting the hint to be within a reasonable range. There should probably be some informative text in the spec about that, even if we don't stipulate the exact bounds normatively.

Yes, the spec should call out that hints that are too close will be set back to 5%

**Comment by yisibl** [comment]
Has this issue been resolved?

**Comment by svgeesus** [comment]
@LeaVerou

**Comment by LeaVerou** [comment]
It is quite unfortunate that this is the behavior we have to settle on, but it's also not realistic to change it at this point. A lesson for everyone about what happens when we underspecify things, lol.

> Has this issue been resolved?

No edits have happened after the May 2019 resolution and from a quick look, the spec still seems to imply that hints do not behave like a regular color stop in a couple places, e.g. here (bold mine): 

> Note: The [transition hint](https://drafts.csswg.org/css-images-4/#color-transition-hint) specifies where the “halfway color”—​the 50% blend between the colors of the two surrounding color stops—​should be placed. **When the hint is exactly halfway between the two surrounding color stops, the above interpolation algorithm happens to produce the ordinary linear interpolation.** If the hint is placed anywhere else, it produces a smooth exponential curve between the surrounding color stops, with the “halfway color” occurring exactly where the hint specifies.
